### DISCUSSION TITLE
Recommended approach for scaling CoreDNS?

### DISCUSSION THREAD

**Discussion by mrparkers** [coredns/coredns#5915] [Q&A] [answered]
Hello!  I'm wondering if there is an official recommendation for the best approach to scale CoreDNS in a large Kubernetes cluster.

After doing a bit of research, a lot of resources online seem to suggest using [cluster-proportional-autoscaler](https://github.com/kubernetes-sigs/cluster-proportional-autoscaler):
- Official Kubernetes docs: https://kubernetes.io/docs/tasks/administer-cluster/dns-horizontal-autoscaling/
- Rancher docs: https://www.suse.com/support/kb/doc/?id=000020133
- GKE docs: https://cloud.google.com/kubernetes-engine/docs/how-to/custom-kube-dns

However, I'm unsure if there is any particular reason why a traditional HPA can't or shouldn't be used here.  The coredns helm chart allows you to configure either an HPA or cluster-proportional-autoscaler, so that isn't much help when trying to understand what the "recommended" approach is.

Does anyone have any thoughts or ideas about this?

Thanks!

**Comment by johnbelamaric** [comment, marked as answer]
Fun question, here are some thoughts.

Cluster proportional autoscaler is used in GKE because GKE uses kube-dns, not CoreDNS. kube-dns consists of multiple containers that have different CPU usage characteristics and so HPA is not really useful for it. As an aside, [this](https://github.com/kubernetes/enhancements/tree/master/keps/sig-autoscaling/1610-container-resource-autoscaling) upcoming K8s feature might help, but it is still not clear to me *which* container would be the limiting factor. GKE also offers using Cloud DNS for the cluster DNS, in which case I think it's scaled for you. See https://cloud.google.com/kubernetes-engine/docs/how-to/cloud-dns

I do not recommend cluster-proportional-autoscaler for CoreDNS. Articles recommending it are probably just carrying forward the old advice from kube-dns. CoreDNS has a single container in the Pod, and is (primarily) CPU-bound. A CPU-based HPA is a better option for CoreDNS.

That said, there are a few subtleties. Depending on a lot of different factors, CoreDNS may saturate the allowable network requests on your node with one or two CPUs. For example, if CoreDNS achieves 25k queries per second per CPU, then with two CPUs you will hit node limits of 50k requests per second from your cloud provider (at least that was a typical limit for AWS years ago); alternatively you may saturate the NIC with enough CPU. There are other, internal factors that can also limit the ability for CoreDNS to fully utilize all the CPU you give it. For example, there is go routine processing overhead, which *seems* to grow non-linearly with number of active routines; there is also #5595.

If you want to be really clever, you can use the metric `coredns_health_request_duration_seconds` (https://github.com/coredns/coredns/tree/master/plugin/health#metrics) for your autoscaling. This is the latency of the health check so is basically directly correlated to how well CoreDNS is able to process requests, taking into account CPU, go routine overhead, and likely the problem discussed in #5595 (not sure on that, it may be HTTP and a different queue). But I am not sure exactly what value to use for autoscaling, or perhaps you need to use some "it's increasing" type of logic.

So, my recommendation would be:
* Set CPU requests and limits to the same thing (guaranteed resources), at least for CPU. Memory is a little harder to predict, it will vary a lot with your query mix, cache configuration, and number of services (and headless endpoints).
* Limit the CPU you request to something that avoids hitting other limits like networking constraints. You'll have to test that in your particular environment. I would start with maybe 1 full CPU.
* Use HPA with a very high CPU percentage, close to 100%. There's no reason to leave CPU on the table. If it's hitting 100% it just means latency goes up a little. The exception would be if you have really spiky queries and tight latency requirements. You may also want to lower the target percentage it if you have lots and lots of services so CoreDNS startup takes a long time. But for most people, having slightly higher latency during scale up is probably safe.
* Use node anti-affinity, if you have NIC or network throttling limits you are getting close to. Two pods on the same node will be throttled just the same as one pod with two CPUs, if the cloud provider is doing a node-level throttle.

You may also want to take a look at @chrisohaver's work from a few years ago:
 * https://coredns.io/2018/11/15/scaling-coredns-in-kubernetes-clusters/ and 
 * https://github.com/coredns/deployment/blob/master/kubernetes/Scaling_CoreDNS.md

**Reply by nmour77** [reply]
@johnbelamaric is there a single autoscaling solution that you would recommend to scale coredns which takes below parameters into consideration? Instead of using different solutions like CPA, HPA and so on. 
- Resource utilization(cpu/memory). 
- Coredns latency. 
- Coredns requests. 
- Number of nodes in the cluster?

**Comment by stp-ip** [comment]
Additional it depends a lot on the specifics of your DNS traffic.
Google does sometime limit the amount of DNS upstream requests per Node (IP), which would also mean that horizontal scaling doesn't help much, if you run into upstream blocked requests.

One "easy" option to make DNS more "scalable" by caching locally and fanning upstream aka non cluster DNS requests to each node is NodeLocal: https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/

**Comment by mrparkers** [comment]
Thank you so much @johnbelamaric and @stp-ip for your helpful answers!  I think I'll go down the route of using an HPA like you suggested, and create an alert for the `coredns_health_request_duration_seconds` metric.

The NodeLocal approach is also really interesting, I may try that a bit later after I've gotten my coredns deployment in a better spot.

Thanks again!
